android/titanium/src/java/org/appcelerator/titanium/io/TiFileFactory.java

```diff
@@ -71,87 +77,94 @@ public static TiBaseFile createTitaniumFile(String path, boolean stream)
-			path = formPath(path, parts);
-			file = new TitaniumBlob(CONTENT_URL_PREFIX + path);
-		} else if (initial.startsWith(ANDROID_RESOURCE_URL_PREFIX)) {
-			// This is an "android.resource://" URL.
-			String path = initial.substring(ANDROID_RESOURCE_URL_PREFIX.length());
-			path = formPath(path, parts);
-			file = new TitaniumBlob(ANDROID_RESOURCE_URL_PREFIX + path);
-		} else if (initial.startsWith("/")) {
-			String path = "";
-			path = formPath(path, insertBefore(path, parts));
-			file = new TiFile(new File(path), FILE_URL_PREFIX + path, stream);
+		String possibleURI = joinPathSegments(parts);
+		String scheme = null;
+		String path = null;
+		int colonIndex = possibleURI.indexOf(':');
+		if (colonIndex != -1) {
+			// probably a URI
+			try {
+				URI uri = new URI(possibleURI);
+				scheme = uri.getScheme().toLowerCase();
+				path = uri.getSchemeSpecificPart();
+			} catch (URISyntaxException use) {
+				// not a valid one!
+				// TODO: Maybe we can encode each segment in joinPathSegments to help avoid this?
+				// hack to grab scheme and path ourselves
+				scheme = possibleURI.substring(0, colonIndex);
+				path = possibleURI.substring(colonIndex + 1);
+			}
+			// if there was a "//" after the scheme, strip it to get path
+			if (path.startsWith("//")) {
+				path = path.substring(2);
+			}
 		} else {
-			String path = "";
-			path = formPath(path, insertBefore(path, parts));
+			// no ':', so no scheme! make the whole thing the path.
+			path = possibleURI;
+			if (path.startsWith("/")) { // absolute path, so assume file:
+				scheme = FILE_URL_SCHEME;
+			} else { // relative looking path, so assume appdata-private:
+				scheme = APPDATA_PRIVATE_URL_SCHEME;
+			}
+		}
+
+		if (TiC.URL_APP_SCHEME.equals(scheme)) {
+			return new TiResourceFile(trimFront(path, '/'));
+		}
+
+		if (APPDATA_PRIVATE_URL_SCHEME.equals(scheme)) {
 			File f = new File(getDataDirectory(true), path);
-			file = new TiFile(f, APPDATA_PRIVATE_URL_PREFIX + path, stream);
+			return new TiFile(f, possibleURI, stream);
 		}
 
-		return file;
-	}
+		if (APPDATA_URL_SCHEME.equals(scheme)) {
+			File f = new File(getDataDirectory(false), path);
+			return new TiFile(f, possibleURI, stream);
+		}
 
-	private static String[] insertBefore(String path, String[] parts)
-	{
-		String[] p = new String[parts.length + 1];
-		p[0] = path;
-		for (int i = 0; i < parts.length; i++) {
-			p[i + 1] = parts[i];
+		if (CONTENT_URL_SCHEME.equals(scheme) || ANDROID_RESOURCE_URL_SCHEME.equals(scheme)) {
+			return new TitaniumBlob(possibleURI); // TODO: Forward along the actual URI instance?
+		}
+
+		if (FILE_URL_SCHEME.equals(scheme)) {
+			// check for fake "file:///android_asset/Resources/" URL, treat like app:
+			if (path.startsWith(ANDROID_ASSET_RESOURCES)) {
+				// Strip this fake base path
+				path = path.substring(ANDROID_ASSET_RESOURCES.length());
+				return new TiResourceFile(trimFront(path, '/')); // remove leading '/' characters
+			}
```

Comment:
The "file:///android_asset/" URL is not a fake path. It's officially supported by Google and is used to access files from the APK's assets folder. But I'm pretty sure it cannot be loaded by a Java `File` class. This seems to be an issue with the old code. We need to do the below else-if block to turn the given URL into a "content://" so that it can be loaded by our own content provider. Note that I haven't tested this yet. I need to confirm that we can access an APK's root "assets" file such as "file:///android_asset/app.json".
```java
} else if (path.startsWith("/android_asset/")) {
	Uri contentUri = TiFileProvider.createUriFrom(possibleURI).toString();
	if (contentUri != null) {
		return new TitaniumBlob(contentUri.toString());
	}
	return null;
}
```

FYI: The "file:///android_asset/Resources" path needs special handling in Titanium because the given path may reference encrypted assets such as a `.js` or `.json` file. This is where a Titanium project's "Resources" files gets dumped.

Reply:
Oh bummer. The above code causes infinite recursion. I'll have to come up with a different solution.

Reply:
Ideally, we should wrap a path such as "file:///android_asset/app.json" with our `TiResourceFile` since it supports fast directory listing support (we can't get a directory listing from a "content://" URL which is what my above code is turning it into). Unfortunately, our `TiResourceFile` class is currently limited to files under the "file:///android_asset/Resources" directory, but I believe @garymathews's upcoming snapshot PR will change this since it now wants the "Resources/" base directory.

So, this should be a task for later.

For now, either return `null` for this case or I guess continue to return a broken `TiFile` like the old code was doing before.